File: apps/ai-service/app/ai/vectorstore.py
from __future__ import annotations

from sqlalchemy import Column, Integer, String, Table, Text, select, text
from sqlalchemy.engine import Engine

from app.ai.embeddings import Embedder
from app.ai.schemas import KBChunk

try:  # pgvector is required at runtime; guarded so imports don't explode in bare envs.
    from pgvector.sqlalchemy import Vector
except ImportError:  # pragma: no cover
    Vector = None  # type: ignore[assignment]


class VectorStore:
    """pgvector-backed knowledge-base store living in the shared Postgres.

    The table is created by an Alembic migration (kb_chunks). This class only
    reads/writes rows and runs semantic (cosine) and keyword (FTS) search.
    """

    def __init__(self, engine: Engine, embedder: Embedder) -> None:
        if Vector is None:  # pragma: no cover
            raise RuntimeError("pgvector is not installed; add 'pgvector' to dependencies")
        self.engine = engine
        self.embedder = embedder
        self.table = Table(
            "kb_chunks",
            _metadata(),
            Column("id", Integer, primary_key=True),
            Column("source", String, nullable=False),
            Column("content", Text, nullable=False),
            Column("embedding", Vector(embedder.dim), nullable=False),
            extend_existing=True,
        )

    def add(self, source: str, content: str) -> None:
        embedding = self.embedder.embed(content)
        with self.engine.begin() as conn:
            conn.execute(
                self.table.insert().values(source=source, content=content, embedding=embedding)
            )

    def search(self, query: str, k: int) -> list[KBChunk]:
        """Backward-compatible alias for semantic search."""
        return self.search_semantic(query, k)

    def search_semantic(self, query: str, k: int) -> list[KBChunk]:
        embedding = self.embedder.embed(query)
        distance = self.table.c.embedding.cosine_distance(embedding)
        stmt = (
            select(
                self.table.c.id,
                self.table.c.content,
                self.table.c.source,
                distance.label("distance"),
            )
            .order_by(distance)
            .limit(k)
        )
        with self.engine.connect() as conn:
            rows = conn.execute(stmt).all()
        return [
            KBChunk(
                id=int(r.id),
                content=r.content,
                source=r.source,
                distance=float(r.distance),
            )
            for r in rows
        ]

    def search_keyword(self, query: str, k: int) -> list[KBChunk]:
        """Postgres full-text search ranked by ``ts_rank_cd``.

        Returns an empty list when the query has no useful terms (so
        ``plainto_tsquery`` would match nothing useful / raise noise).
        Distance is set to ``1 - rank`` (clipped) so lower-is-better still holds.
        """
        q = (query or "").strip()
        if not q:
            return []

        # Use a single SQL statement so the GIN expression index can be used.
        # Rank is higher-is-better; we invert for KBChunk.distance consistency.
        sql = text(
            """
            SELECT id, content, source,
                   ts_rank_cd(to_tsvector('english', content), query) AS rank
            FROM kb_chunks,
                 plainto_tsquery('english', :q) AS query
            WHERE to_tsvector('english', content) @@ query
            ORDER BY rank DESC
            LIMIT :k
            """
        )
        with self.engine.connect() as conn:
            rows = conn.execute(sql, {"q": q, "k": k}).all()
        return [
            KBChunk(
                id=int(r.id),
                content=r.content,
                source=r.source,
                distance=max(0.0, 1.0 - float(r.rank)),
            )
            for r in rows
        ]


def _metadata():
    from sqlalchemy import MetaData

    return MetaData()

File: apps/ai-service/internal/rag/vectorstore.go
package rag

import (
	"context"
	"database/sql"
	"regexp"
	"strings"

	"github.com/pgvector/pgvector-go"
)

// VectorStore is the pgvector-backed knowledge-base store living in the shared
// Postgres.
//
// The kb_chunks table is created by this service's migrations. This type only
// reads/writes rows and runs semantic (cosine) and keyword (FTS) search.
type VectorStore struct {
	db       *sql.DB
	embedder Embedder
}

// NewVectorStore wires a store to a database handle and an embedder. The
// embedder's Dim must match the kb_chunks.embedding column width.
func NewVectorStore(db *sql.DB, embedder Embedder) *VectorStore {
	return &VectorStore{db: db, embedder: embedder}
}

// Embedder exposes the configured embedder (ingestion reuses it).
func (s *VectorStore) Embedder() Embedder { return s.embedder }

// Add embeds content and stores it as a knowledge-base chunk.
func (s *VectorStore) Add(ctx context.Context, source, content string) error {
	embedding, err := s.embedder.Embed(ctx, content)
	if err != nil {
		return err
	}
	_, err = s.db.ExecContext(ctx,
		`INSERT INTO kb_chunks (source, content, embedding) VALUES ($1, $2, $3)`,
		source, content, pgvector.NewVector(embedding))
	return err
}

// Search is a backward-compatible alias for semantic search.
func (s *VectorStore) Search(ctx context.Context, query string, k int) ([]KBChunk, error) {
	return s.SearchSemantic(ctx, query, k)
}

// SearchSemantic ranks chunks by cosine distance to the query embedding.
func (s *VectorStore) SearchSemantic(ctx context.Context, query string, k int) ([]KBChunk, error) {
	embedding, err := s.embedder.Embed(ctx, query)
	if err != nil {
		return nil, err
	}
	const stmt = `
		SELECT id, content, source, embedding <=> $1 AS distance
		FROM kb_chunks
		ORDER BY embedding <=> $1
		LIMIT $2`
	rows, err := s.db.QueryContext(ctx, stmt, pgvector.NewVector(embedding), k)
	if err != nil {
		return nil, err
	}
	return scanChunks(rows)
}

// SearchKeyword runs Postgres full-text search ranked by ts_rank_cd.
//
// Terms are OR-ed: plainto_tsquery ANDs every lexeme, so a multi-sentence
// ticket (title + description) matched nothing unless a single chunk contained
// *all* its words. OR-ing lets ts_rank_cd reward chunks that hit the most terms
// instead.
//
// Returns no rows when the query has no useful terms. Distance is set to
// 1 - rank (clipped at 0) so lower-is-better still holds.
func (s *VectorStore) SearchKeyword(ctx context.Context, query string, k int) ([]KBChunk, error) {
	terms := tsqueryOrTerms(query)
	if terms == "" {
		return nil, nil
	}
	// A single statement so the GIN expression index can be used. Rank is
	// higher-is-better; we invert it for KBChunk.Distance consistency.
	const stmt = `
		SELECT id, content, source,
		       1 - ts_rank_cd(to_tsvector('english', content), query) AS distance
		FROM kb_chunks,
		     to_tsquery('english', $1) AS query
		WHERE to_tsvector('english', content) @@ query
		ORDER BY ts_rank_cd(to_tsvector('english', content), query) DESC
		LIMIT $2`
	rows, err := s.db.QueryContext(ctx, stmt, terms, k)
	if err != nil {
		return nil, err
	}
	chunks, err := scanChunks(rows)
	if err != nil {
		return nil, err
	}
	for i := range chunks {
		if chunks[i].Distance < 0 {
			chunks[i].Distance = 0
		}
	}
	return chunks, nil
}

func scanChunks(rows *sql.Rows) ([]KBChunk, error) {
	defer rows.Close()
	var out []KBChunk
	for rows.Next() {
		var c KBChunk
		if err := rows.Scan(&c.ID, &c.Content, &c.Source, &c.Distance); err != nil {
			return nil, err
		}
		out = append(out, c)
	}
	return out, rows.Err()
}

var termRE = regexp.MustCompile(`[a-z0-9]+`)

// tsqueryOrTerms builds an OR-ed to_tsquery input from free text.
//
// Tokens are restricted to alphanumerics so user text can't inject tsquery
// syntax; duplicates are dropped (order-preserving). Stopwords are left for
// to_tsquery('english', ...) to discard.
func tsqueryOrTerms(query string) string {
	seen := make(map[string]struct{})
	var terms []string
	for _, token := range termRE.FindAllString(strings.ToLower(query), -1) {
		if _, ok := seen[token]; ok {
			continue
		}
		seen[token] = struct{}{}
		terms = append(terms, token)
	}
	return strings.Join(terms, " | ")
}
